src/core/symbology/qgssymbol.cpp

```diff
@@ -81,6 +81,85 @@ QgsSymbol::QgsSymbol( Qgis::SymbolType type, const QgsSymbolLayerList &layers )
 Q_NOWARN_DEPRECATED_POP
 
 QPolygonF QgsSymbol::_getLineString( QgsRenderContext &context, const QgsCurve &curve, bool clipToExtent )
+{
+  if ( curve.is3D() )
+    return _getLineString3d( context, curve, clipToExtent );
+  else
+    return _getLineString2d( context, curve, clipToExtent );
+}
+
+QPolygonF QgsSymbol::_getLineString3d( QgsRenderContext &context, const QgsCurve &curve, bool clipToExtent )
+{
+  const unsigned int nPoints = curve.numPoints();
+
+  QgsCoordinateTransform ct = context.coordinateTransform();
+  const QgsMapToPixel &mtp = context.mapToPixel();
+  QgsLineString pts;
+
+  //apply clipping for large lines to achieve a better rendering performance
+  if ( clipToExtent && nPoints > 1 && !( context.flags() & QgsRenderContext::ApplyClipAfterReprojection ) )
+  {
+    const QgsRectangle e = context.extent();
+    const double cw = e.width() / 10;
+    const double ch = e.height() / 10;
+    const QgsBox3d clipRect( e.xMinimum() - cw, e.yMinimum() - ch, -HUGE_VAL, e.xMaximum() + cw, e.yMaximum() + ch, HUGE_VAL ); // TODO also need to be clipped according to z axis
+    pts = QgsClipper::clipped3dLine( curve, clipRect );
+  }
+  else
+  {
+    // clone...
+    QgsPointSequence seq;
```

Comment:
Thanks for separating these code paths. Just keep in mind that performance with 3d lines will still be quite poor until you avoid all these container conversions and move to raw double lists instead...

Reply:
@nyalldawson 

By the way, I added a new function to QgsLineString in order to use the double * data : `pts.setPoints( tmpLine->xData(), tmpLine->yData(), tmpLine->zData(), tmpLine->mData(), curve.numPoints() );`. Do you prefer this solution?

Reply:
@nyalldawson If the solution above suit you, I can move the prototype to QgsCurve and implements it into the other sbuclasses?